### PR TITLE
Heterogenous Freezing

### PR DESCRIPTION
**Pull request by pressel**
This is a work in progress PR that implements heterogenous freezing in mam4xx.

### CODE REVIEW

**Comment by pressel** [comment]
I'm moving this PR out of WIP status. Seems validation tests are failing only on deception@PNNL, which seems like a compilation error of some sort. @CameronRutherford do you mind taking a look at what is going on there.

File: src/tests/mam4_test_config.hpp [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#ifndef MAM4_TEST_CONFIG_HPP
+#define MAM4_TEST_CONFIG_HPP
+
+// clang-format off
+#define MAM4_TEST_DATA_DIR "/Users/pres026/Research/EAGLES/mam4xx/src/tests/data"
```
Comment (jeff-cohere) [inline src/tests/mam4_test_config.hpp:5]: Is this used? I bet this directory isn't going to be present in most environments.

File: src/validation/CMakeLists.txt [hunk 2]
```diff
@@ -1,18 +1,18 @@
...
+#views, atmospheric state, prognostics, and diagnostics used by unit tests.
 add_library(validation validation.cpp ${CMAKE_CURRENT_SOURCE_DIR}/../tests/testing.cpp)
 target_link_libraries(validation skywalker)
 include_directories(${CMAKE_CURRENT_SOURCE_DIR})
 include_directories(${PROJECT_BINARY_DIR}/include) # for skywalker
 
-#-------------------------
-# MAM4xx Skywalker drivers
-#-------------------------
+#-- -- -- -- -- -- -- -- -- -- -- -- -
```
Comment (jeff-cohere) [inline src/validation/CMakeLists.txt:13]: Is there a reason you've removing spaces after `#` in this file?

File: src/validation/hetfrz/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+# Copy some Python scripts from mam_x_validation to our binary directory.
+ configure_file(
+     ${HETFRZ_VALIDATION_SCRIPTS_DIR}/compare_hetfzr.py
+     ${CMAKE_CURRENT_BINARY_DIR}/compare_hetfzr.py
+     COPYONLY
+   )
+
+# Run the driver in several configurations to produce datasets.
+
+foreach (input
```
Comment (jeff-cohere) [inline src/validation/hetfrz/CMakeLists.txt:46]: Wow!

File: src/tests/mam4_test_config.hpp (hunk 1 above)
Comment (pressel) [reply src/tests/mam4_test_config.hpp:5]: Oh strange, that looks like IDE funny business.

File: src/validation/hetfrz/hetfrz_calculate_hetfrz_deposition_nucleation.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    auto r_dust_a1 = input.get("r_dust_a1");
+    auto r_dust_a3 = input.get("r_dust_a3");
+    auto do_bc = input.get("do_bc");
+    auto do_dst1 = input.get("do_dst1");
+    auto do_dst3 = input.get("do_dst3");
+
+    bool do_bc_b = true, do_dst1_b = true, do_dst3_b = true;
+
+    // Convert real to bool
+    if (do_bc != 0.0) {
```
Comment (jeff-cohere) [inline src/validation/hetfrz/hetfrz_calculate_hetfrz_deposition_nucleation.cpp:114]: It seems to me that these expressions should be reversed. E.g. `if (do_bc != 0.0) { do_bc_b = true; }`.

File: src/validation/hetfrz/hetfrz_calculate_hetfrz_immersion_nucleation.cpp [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    auto rgimm_dust_a1 = input.get("rgimm_dust_a1");
+    auto rgimm_dust_a3 = input.get("rgimm_dust_a3");
+    auto do_bc = input.get("do_bc");
+    auto do_dst1 = input.get("do_dst1");
+    auto do_dst3 = input.get("do_dst3");
+
+    bool do_bc_b = true, do_dst1_b = true, do_dst3_b = true;
+
+    // Convert real to bool
+    if (do_bc != 0.0) {
```
Comment (jeff-cohere) [inline src/validation/hetfrz/hetfrz_calculate_hetfrz_immersion_nucleation.cpp:156]: These seem backwards too.

File: src/validation/hetfrz/hetfrz_calculate_rgimm_and_determine_spec_flag.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    auto aw = input.get("aw");
+    auto supersatice = input.get("supersatice");
+    auto rgimm = input.get("rgimm");
+
+    bool do_spec_flag = true;
+    hetfrz::calculate_rgimm_and_determine_spec_flag(
+        vwice, sigma_iw, temperature, aw, supersatice, rgimm, do_spec_flag);
+
+    skywalker::Real do_spec_flag_real;
+    if (do_spec_flag) {
```
Comment (jeff-cohere) [inline src/validation/hetfrz/hetfrz_calculate_rgimm_and_determine_spec_flag.cpp:65]: These also seem backwards, in the sense that a non-zero value is usually interpreted as `true` and a zero value as `false`. But maybe you've adopted the opposite convention.

File: src/tests/mam4_test_config.hpp (hunk 1 above)
Comment (pressel) [reply src/tests/mam4_test_config.hpp]: Removed.

File: src/mam4xx/hetfrz.hpp [hunk 7]
```diff
@@ -0,0 +1,1490 @@
...
+    Real uncoated_aer_num[Hetfrz::hetfrz_aer_nspec], const Real icnlx,
+    const Real sigma_iv, const Real eswtr, const Real rgimm, const Real r_bc,
+    const Real r_dust_a1, const Real r_dust_a3, const Real Kcoll_bc,
+    const Real Kcoll_dust_a1, const Real Kcoll_dust_a3, const bool do_bc,
+    const bool do_dst1, const bool do_dst3, Real &frzbccnt, Real &frzducnt) {
+
+  frzbccnt = 0.0;
+  frzducnt = 0.0;
+
+  constexpr Real bad_pi = 3.141592654; // (BAD CONSTANT)
```
Comment (jeff-cohere) [inline src/mam4xx/hetfrz.hpp:338]: We have at least 2 pi in this file!

File: src/mam4xx/hetfrz.hpp [hunk 8]
```diff
@@ -0,0 +1,1490 @@
...
+      get_form_factor(Hetfrz::theta_dep_dust * bad_pi / 180.0);
+  const Real f_cnt_dust_a3 =
+      get_form_factor(Hetfrz::theta_dep_dust * bad_pi / 180.0);
+
+  // homogeneous energy of germ formation
+  const Real dg0cnt = 4.0 * bad_pi / 3.0 * sigma_iv * haero::square(rgimm);
+
+  // prefactor
+  // attention: division of small numbers
+  constexpr Real bad_boltzmann = 1.38e-23; // Boltzmann Constant (BAD CONSTANT)
```
Comment (jeff-cohere) [inline src/mam4xx/hetfrz.hpp:352]: Good grief.

File: src/mam4xx/hetfrz.hpp [hunk 9]
```diff
@@ -0,0 +1,1490 @@
...
+                           Ktherm_dust, Pr, Tdiff, Kcoll_dust_a1);
+
+  calculate_collkernel_sub(temperature, pressure, rho_air, r3lx, r_dust_a3,
+                           lambda, latvap, viscos_air, Re, Ktherm_air,
+                           Ktherm_dust, Pr, Tdiff, Kcoll_dust_a3);
+}
+
+KOKKOS_INLINE_FUNCTION
+Real get_form_factor(const Real alpha) {
+  const Real v_cos = cos(alpha); // This should resolve to haero cos
```
Comment (jeff-cohere) [inline src/mam4xx/hetfrz.hpp:302]: You could sidestep this comment and just use `haero::cos` if you like.

Comment (pressel) [reply src/mam4xx/hetfrz.hpp:302]: Ah, yes, I think I actually added `haero::cos` and then forgot to use it.

**Review by jeff-cohere — APPROVED** [review]
Well done, sir. I left a few comments here and there, the most significant having to do with the reversed (but consistent) convention for interpreting `Real`s as `bool`s and vice-versa. Up to you if you want to adjust that. I think this is good to merge when we figure out what's going on with Deception and after you and other reviewers are satisfied with everything.

**Comment by CameronRutherford** [comment]
> I'm moving this PR out of WIP status. Seems validation tests are failing only on deception@PNNL, and it seems like a compilation error of some sort associated with convective aerosol processes that are not touched by this PR. @CameronRutherford do you mind taking a look at what is going on there.

Hello, it is I, the CI magician who can read logs really well. Looks like you have a std::cout in your device code from reading the logs https://code.pnnl.gov/e3sm/eagles/mam4xx/-/jobs/14337#L1021

File: src/mam4xx/hetfrz.hpp [hunk 10]
```diff
@@ -0,0 +1,1490 @@
...
+  auto &coarse_mom = progs.q_aero_i[coarse_idx][int(AeroId::MOM)];
+  auto &coarse_bc = progs.q_aero_i[coarse_idx][int(AeroId::BC)];
+  auto &coarse_pom = progs.q_aero_i[coarse_idx][int(AeroId::POM)];
+  auto &coarse_soa = progs.q_aero_i[coarse_idx][int(AeroId::SOA)];
+  auto &coarse_so4 = progs.q_aero_i[coarse_idx][int(AeroId::SO4)];
+
+  auto &num_coarse = progs.n_mode_i[coarse_idx];
+
+  // initialize rho
+  std::cout.precision(17);
```
Comment (CameronRutherford) [inline src/mam4xx/hetfrz.hpp:1253]: @pressel here is your CI bug. Only happens on double precision for some reason, but a valid failure

File: src/mam4xx/hetfrz.hpp [hunk 11]
```diff
@@ -0,0 +1,1490 @@
...
+  void compute_tendencies(const AeroConfig &config, const ThreadTeam &team,
+                          Real t, Real dt, const Atmosphere &atm,
+                          const Prognostics &progs, const Diagnostics &diags,
+                          const Tendencies &tends) const;
+};
+
+namespace hetfrz {
+
+KOKKOS_INLINE_FUNCTION
+Real get_air_viscosity(const Real tc) {
```
Comment (odiazib) [inline src/mam4xx/hetfrz.hpp:119]: Is this subroutine used by other process? if so, I would move it to another file.  Also, we also have another implementation of dynamical viscosity of air [here](https://github.com/eagles-project/mam4xx/blob/f5c775918f620a9b3ffa29620ca50c1b86fc0019/src/mam4xx/aero_model.hpp#L113); however, the correlation are different.

File: src/mam4xx/hetfrz.hpp [hunk 12]
```diff
@@ -0,0 +1,1490 @@
...
+  const Real Daer =
+      kboltz2 * temperature * (1 + Kn) / (6.0 * pi * r_a * viscos_air);
+
+  // Schmidt number
+  const Real Sc = viscos_air / (Daer * rho_air);
+  //  Young (1974) first equ. on page 771
+  // Todo (the 0.33 power here should probably be 1/3, but that wouldn't be
+  // BFB.)
+
+  const Real K_brownian = 4.0 * pi * r3lx * Daer *
```
Comment (odiazib) [inline src/mam4xx/hetfrz.hpp:225]: I would change 0.33 to 1./3.

File: src/mam4xx/hetfrz.hpp [hunk 13]
```diff
@@ -0,0 +1,1490 @@
...
+                          (1.0 + 0.3 * haero::sqrt(Re) * haero::pow(Sc, 0.33));
+
+  // form factor
+  const Real f_t =
+      0.4 * (1.0 + 1.45 * Kn + 0.4 * Kn * haero::exp(-1. / Kn)) *
+      (Ktherm_air + 2.5 * Kn * Ktherm) /
+      ((1.0 + 3.0 * Kn) * (2.0 * Ktherm_air + 5.0 * Kn * Ktherm + Ktherm));
+  const Real Q_heat = Ktherm_air / r3lx *
```
Comment (odiazib) [inline src/mam4xx/hetfrz.hpp:233]: I would change 0.33 to 1./3.

File: src/mam4xx/hetfrz.hpp [hunk 14]
```diff
@@ -0,0 +1,1490 @@
...
+}
+
+KOKKOS_INLINE_FUNCTION
+void collkernel(const Real temperature, const Real pressure, const Real eswtr,
+                const Real r3lx, const Real r_bc, const Real r_dust_a1,
+                const Real r_dust_a3, Real &Kcoll_bc, Real &Kcoll_dust_a1,
+                Real &Kcoll_dust_a3) {
+
+  // thermal conductivities from Seinfeld & Pandis, Table 8.6
+  constexpr Real Ktherm_bc = 4.2;    // Carbon
```
Comment (odiazib) [inline src/mam4xx/hetfrz.hpp:253]: BAD CONSTANT?

File: src/mam4xx/hetfrz.hpp [hunk 15]
```diff
@@ -0,0 +1,1490 @@
...
+  constexpr Real Ktherm_dust = 0.72; // Clay
+
+  const Real tc = temperature - Constants::freezing_pt_h2o;
+
+  // air viscosity for tc<0,
+  const Real viscos_air = get_air_viscosity(tc);
+
+  const Real r_gas_dry_air = 287.04231136504870; // (BAD CONSTANT)
```
Comment (odiazib) [inline src/mam4xx/hetfrz.hpp:261]: Use constant from haero::Constant

File: src/mam4xx/hetfrz.hpp [hunk 16]
```diff
@@ -0,0 +1,1490 @@
...
+Real get_dg0imm(const Real sigma_iw, const Real rgimm) {
+  return 4.0 * Constants::pi / 3.0 * sigma_iw * haero::square(rgimm);
+}
+
+KOKKOS_INLINE_FUNCTION
+Real get_Aimm(const Real vwice, const Real rgimm, const Real temperature,
+              const Real dg0imm) {
+  constexpr Real n1 = 1e19; // number of water molecules in contact with unit
+                            // area of substrate [m-2]
+  constexpr Real hplanck = 6.63e-34; // Planck constant (BAD CONSTANT)
```
Comment (odiazib) [inline src/mam4xx/hetfrz.hpp:316]: move constant to haero::Constant.

File: src/mam4xx/hetfrz.hpp [hunk 17]
```diff
@@ -0,0 +1,1490 @@
...
+
+KOKKOS_INLINE_FUNCTION
+Real get_dg0imm(const Real sigma_iw, const Real rgimm) {
+  return 4.0 * Constants::pi / 3.0 * sigma_iw * haero::square(rgimm);
+}
+
+KOKKOS_INLINE_FUNCTION
+Real get_Aimm(const Real vwice, const Real rgimm, const Real temperature,
+              const Real dg0imm) {
+  constexpr Real n1 = 1e19; // number of water molecules in contact with unit
```
Comment (odiazib) [inline src/mam4xx/hetfrz.hpp:314]: BAD CONSTANT?

File: src/mam4xx/hetfrz.hpp [hunk 18]
```diff
@@ -0,0 +1,1490 @@
...
+  const Real f_cnt_dust_a3 =
+      get_form_factor(Hetfrz::theta_dep_dust * bad_pi / 180.0);
+
+  // homogeneous energy of germ formation
+  const Real dg0cnt = 4.0 * bad_pi / 3.0 * sigma_iv * haero::square(rgimm);
+
+  // prefactor
+  // attention: division of small numbers
+  constexpr Real bad_boltzmann = 1.38e-23; // Boltzmann Constant (BAD CONSTANT)
+  constexpr Real mwh2o =
```
Comment (odiazib) [inline src/mam4xx/hetfrz.hpp:353]: Use constant from haero.

File: src/mam4xx/hetfrz.hpp [hunk 19]
```diff
@@ -0,0 +1,1490 @@
...
+
+  const Real bc_kg_to_num = 4.669152e+17;
+  coated_aer_num[0] =
+      bcmpc * bc_kg_to_num * Hetfrz::num_m3_to_cm3 * dstcoat[0] +
+      bcmac * bc_kg_to_num * Hetfrz::num_m3_to_cm3;
+
+  uncoated_aer_num[0] =
+      bcmpc * bc_kg_to_num * Hetfrz::num_m3_to_cm3 * (1.0 - dstcoat[0]);
+
+  const Real dst1_0p5um_scale = 0.488;
```
Comment (odiazib) [inline src/mam4xx/hetfrz.hpp:1056]: BAD CONSTANT?

**Review by odiazib — APPROVED** [review]
Nice work! I added a few comments about constants.

File: src/mam4xx/hetfrz.hpp (hunk 10 above)
Comment (pressel) [reply src/mam4xx/hetfrz.hpp:1253]: Thanks for tracking this down @CameronRutherford, somehow I missed it when looking through the CI Logs.

File: src/validation/hetfrz/CMakeLists.txt (hunk 3 above)
Comment (pressel) [reply src/validation/hetfrz/CMakeLists.txt:46]: Too much validation yet?

File: src/validation/CMakeLists.txt (hunk 2 above)
Comment (pressel) [reply src/validation/CMakeLists.txt:13]: No idea how or why that happened, I'll fix it.

File: src/mam4xx/hetfrz.hpp (hunk 12 above)
Comment (jeff-cohere) [reply src/mam4xx/hetfrz.hpp:225]: BFB is nice to have at this particular moment, so maybe we can add this to a GitHub issue to revisit later. I don't remember whether we have a "things to go back and fix" issue (I think not), but we could start anytime!

**Comment by jeff-cohere** [comment]
@pressel , I don't have strong feelings about test coverage, but if you feel there's anything that needs better coverage before this goes in, I'd be interested to hear it.

File: src/mam4xx/hetfrz.hpp (hunk 12 above)
Comment (pressel) [reply src/mam4xx/hetfrz.hpp:225]: This is a direct port from F90, so switching from 0.33 to 1.0/.30 or using `haero::cbrt()` would change the solution.

File: src/mam4xx/hetfrz.hpp (hunk 13 above)
Comment (pressel) [reply src/mam4xx/hetfrz.hpp:233]: This would no longer be a direct port.

**Comment by pressel** [comment]
> 

@overfelt sent me some good suggestions on how to increase the code-coverage in the validation tests, but practically this doesn't really increase the actual coverage of the ported code. I'm going to make an issue reminding us to use these suggestions from @overfelt , as they could/should be applied more broadly than just heterogeneous freezing.

**Review by overfelt — APPROVED** [review]

**Review by overfelt — APPROVED** [review]

**Review by overfelt — APPROVED** [review]

**Review by overfelt — APPROVED** [review]

**Comment by jeff-cohere** [comment]
@singhbalwinder and @CameronRutherford: @pressel and I would like to get this PR merged today if possible, so if you have time to take a look today, we'd appreciate your input. Thank you!

**Pull request merged**